### DISCUSSION TITLE
message instance retrieval / PSR HTTP message in VerificationRequest

### DISCUSSION THREAD

**Discussion by doganoo** [netflie/whatsapp-cloud-api#136] [Ideas]
thank you for this great project!

I have two suggestions/RFC's and would like to find out if this is supported.

1. public getter for `$message` in `Netflie\WhatsAppCloudApi\Request\MessageRequest`:

```
    public function getMessage(): Message {
        return $this->message;
    }

```

Background: retrieving the message instance enables to log what is sent to the API.

2. letting `Netflie\WhatsAppCloudApi\WebHook\VerificationRequest` return `Psr\Http\Message\ResponseInterface` instead of emitting the HTTP codes.

```
public function validate(array $payload): ResponseInterface
    {
        $mode = $payload['hub_mode'] ?? null;
        $token = $payload['hub_verify_token'] ?? null;
        $challenge = $payload['hub_challenge'] ?? '';

        if ('subscribe' !== $mode || $token !== $this->verify_token) {
            return new EmptyResponse(403);
        }

        return new TextResponse($challenge, 200);
    }
```

the `EmptyResponse` and `TextResponse` classes have to be implemented.

I would be ready to implement this if it is welcome. Please let me know what you think.

**Comment by aalbarca** [comment]
Hi @doganoo your suggestions are always welcome.

1. I'm not sure. I think this would weaken the encapsulation of the class and the benefits are unclear.

I think that clients what consume this library should be able to write logs of their messages without knowing internal details of how the library is built or its objects.

2. It seems like a fantastic idea to me, but modifying the signature of a method would cause a break in the current major version and this would force a major version to be released. But we still do not have significant changes that justify this version jump.

Maybe I could create a dev-3.0 branch and integrate the change there.

**Comment by doganoo** [comment]
Hi @aalbarca 

sorry for my late reply and thank you for your input.

1. the reason from my side is to have the opportunity to log the message content and some metadata. 

> I think that clients what consume this library should be able to write logs of their messages without knowing internal details of how the library is built or its objects.

I see the point with encapsulation, but do you see any other options? for instance, can we implement an event handler or something similar that gives us the message once sent/received?

2. I see the point with the BC. What do you think about creating a new class and declaring the old one as deprecated? This way you are not forced to create a new version yet :)

As I said, I am open to contribute once you agree :)
